### DISCUSSION TITLE
CAP란?

### DISCUSSION THREAD

**Discussion by GIVEN53** [2024-woowacourse-study/level-interview#193] [Database]
참고 자료입니다

https://www.ibm.com/kr-ko/topics/cap-theorem
https://f-lab.kr/insight/database-selection-guide

![image](https://github.com/2024-woowacourse-study/level-interview/assets/64410384/44f52e69-f5f3-4c45-8ad7-24c564b0eeca)

**Comment by nayonsoso** [comment]
### CAP 이론이란?
분산된 데이터베이스 시스템은 일관성(Consistency), 가용성(Availability), 분할 내성(Partition Tolerance)의 3가지 특성 중 2가지 특성만을 충족할 수 있고 3가지를 모두 충족할 수는 없다는 이론이다. 
즉, CA / AP / CP 만 가능하다는 뜻.

주의할 것은, 이는 ‘네트워크로 연결된 분산 데이터 시스템’에 해당하는 내용이라는 것이다. 
로컬에서 DB를 사용할 때는 적용되지 않는다.

### 분산 시스템이란?
분산 시스템은 동시에 둘 이상의 노드에 데이터를 저장하는 네트워크이다. 
모든 클라우드 애플리케이션은 분산 시스템이므로, 
애플리케이션에 가장 필요한 특성을 제공하는 데이터 관리 시스템을 선택하려면 클라우드 앱의 설계 시 CAP 정리를 반드시 이해하고 있어야 한다.

### 분산 데이터베이스 시스템이 왜 필요한가?
1. 고가용성 : 한 노드가 다운되더라도 어플리케이션이 계속 작동할 수 있다.
2. 지리적 접근성 향상 : 사용자의 지리적 위치에 따라 더 빠른 응답을 할 수 있다. 
3. 트래픽 분산(로드 밸런싱) : 트래픽을 균등하게 분산해서 특정 서버에 부하가 걸리는 것을 막을 수 있다.

---

### CAP의 각 요소

#### 일관성(Consistency)
일관성이란, “모든 클라이언트가 동시에 동일 데이터를 볼 수 있음”을 의미한다. 
즉, 모든 읽기에 대해서 DB노드가 항상 동일한 데이터를 가지고 있어야한다.
이를 위해서는 데이터가 하나의 노드에 기록될 때마다, 이것이 성공으로 간주되기 전에, 즉시 다른 모든 노드로 동기화되어야 한다.

#### 가용성(Availability)
가용성이란, “하나 이상의 노드가 작동 중지된 경우에도 데이터를 요청하는 클라이언트가 응답을 받을 수 있음”을 의미한다. 
즉, 일부 노드에 장애가 생기더라도 다른 작업 중인 노드가 요청에 대한 유효한 응답을 한다.

#### 분할 내성(Partition Tolerance)
분할이란, 두 노드 간의 연결이 유실되거나 일시적으로 지연된 상태이다. 
분할 내성이란, 시스템의 “노드 간에 통신 단절에도 불구하고 각 노드가 계속해서 작동함”을 의미한다.
예를 들어, A 와 B 라는 노드에서 B에 장애가 발생했다면, A는 B의 상태를 몰라도 A 자체적으로 동작하는 것이다.

---

### CP / AP / CA

#### CP 데이터베이스
CP 데이터베이스는 가용성을 희생시키면서 일관성과 분할 내성을 제공한다. 
두 노드 간에 분할이 발생하면, 시스템은 분할이 해결될 때까지 일관되지 않은 노드를 종료한다.

#### AP 데이터베이스
AP 데이터베이스는 일관성을 희생시키면서 가용성과 분할 내성을 제공한다. 
분할이 발생했을 때 모든 노드가 사용 가능한 상태로 유지되지만, 이들이 동기화되지 않을 수 있기 때문에 일관되지 않은 데이터를 리턴할 수 있다.
(분할이 해결되면 AP 데이터베이스는 일반적으로 노드를 재동기화하여 시스템의 모든 불일치를 복구한다.)

#### CA 데이터베이스
CA 데이터베이스는 모든 노드에서 일관성과 가용성을 제공한다. 
그러나 분산 시스템에서는 분할을 피할 수 없고, 분할이 된 상태에서는 일관성을 제공할 수 없다.
따라서 CA 분산 데이터베이스는 이론적으로 논의할 수 있지만, 
현실에서 분할 내성을 희생하며 일관성&가용성을 제공하는 DB는 존재할 수 없다. 
(IBM 피셜 : https://www.ibm.com/kr-ko/topics/cap-theorem)

---

### CAP를 기준으로 DB를 분류..?
CAP 이론은 분산 시스템을 이해하고 설계하는 데 중요한 틀을 제공하지만, 
최근의 데이터베이스 기술 발전으로 인해 이 이론을 따르는 것만으로는 모든 데이터베이스 시스템의 다양성과 유연성을 완전히 설명하기 어려워졌다. 
이는 특히 데이터베이스 시스템이 수직적 구조(예: 마스터-슬레이브 구조)에서 수평적 구조(예: 클러스터 구성)로 진화하면서 더욱 명확해졌다.

e.g. master-slave 구조를 갖는 MySQL의 경우 대표적인 CA 데이터베이스로 분류되곤 했다.
master-slave	구조에서는 마스터 노드가 쓰기 작업을 처리하고 슬레이브 노드가 주로 읽기 작업을 처리하기 때문에
일관성(Consistency)와 가용성(Availability)을 제공한다.
그러나, 마스터 노드에 대한 슬레이브 노드의 의존이 높기 때문에 분할 내성은 제공하지 않았다.
하지만, MySQL 도 클러스터 설정(여러 노드가 동등하게 데이터를 처리하고 서로 동기화를 유지하는 구조)이 가능하기 때문에 
일관성과 분할 내성(Partition Tolerance)을 제공하는 데이터베이스가 될 수 있다.

### 정리
CAP 이론은 그 자체로 이론적인 한계가 있으며, (P에 대한 정의가 명확하지 않다는 지적이 있음)
CAP 이론을 데이터베이스 시스템 분류의 유일한 기준으로 사용하는 것은 요즘 DB의 복잡성과 유연성을 모두 설명하지 못한다.
이러한 문제점을 극복하고자 PACELC 이론이 나오기도 했다. 
(PACELC 이론에서는 분할의 상황과 그렇지 않은 상황을 구분한다)

그럼에도 불구하고, CAP의 각 기준들을 고려하는 것은 여전히 장점을 갖는다. 
CAP의 각 기준은 분산 데이터베이스의 기본적인 이해를 돕고, 시스템 설계 초기 단계에서 중요한 의사 결정을 하는 데 도움을 줄 수 있다. 
또한 데이터 일관성, 가용성, 분할 내성 간의 균형을 이해하고 조정하는 데 필요한 틀을 제공한다.

**Comment by gyungchan-jo** [comment]
CAP 이론에 따르면, 분산 시스템은 세 가지 속성(일관성-C, 가용성-A, 분할 내성-P) 중 두 가지만 선택할 수 있습니다. 
따라서, 시스템 설계 시 어떤 속성을 우선시할 것인지 결정해야 합니다.

1. CP (Consistency + Partition Tolerance):
- 특징: 시스템은 일관성을 유지하면서 네트워크 분할을 허용합니다. (데이터 정확성이 중요한 경우)
- 예시: Redis, MongoDB
- 사용 사례: 금융 거래 시스템, 은행 시스템 등 일관성이 중요한 시스템

2. AP (Availability + Partition Tolerance):
- 특징: 시스템은 가용성을 유지하면서 네트워크 분할을 허용합니다. (서비스 지속성이 중요한 경우)
- 예시: Cassandra, DynamoDB
- 사용 사례: 소셜 미디어, 웹 애플리케이션 등 가용성이 중요한 시스템
